### DISCUSSION TITLE
Trying to create an object on my device

### DISCUSSION THREAD

**Discussion by tnordstrom** [ChristianTremblay/BAC0#437] [Q&A]
I'm trying to create an object on my device (not create a local object), I'm using syntax:

bacnet.create_object(analog_value, device1, "TEST1", "Did this just get created?", presentValue=1.0, commandable=True)

But when this runs I get error AttributeError: 'Lite' object has no attribute 'create_object'.

I don't understand, I'm not starting in 'Lite' mode, I use this:

bacnet = BAC0.connect(ip='192.168.17.186/21')

Anyone with any insight? Appreciation in advance.

**Comment by bbartling** [comment]
if you want to make a BACnet server maybe look at this as an example, its just a bunch of random points and simulate values.
https://github.com/ChristianTremblay/BAC0/issues/381#issue-1607475583

**Reply by tnordstrom** [reply]
I don't want to create a server, I want to create an object on my real device.

**Comment by bbartling** [comment]
so if you only want a BACnet client device to just read and write to other devices...
https://bac0.readthedocs.io/en/latest/read.html#read-examples

This is a basic example of just making simple read requests.
https://gist.github.com/bbartling/88e95557b0b297bea721c0f285657d53#file-simple_read_request-py

To me when someone states "objects" at least in my mind its affiliated with a BACnet server project where a client device is requesting some sort of data associated with an "object" like a point value or name or device description.... where then the client script requests whatever BACnet object that the server is hosting. I could be wrong with the terminology but that's my best guess...

**Reply by tnordstrom** [reply]
https://bac0.readthedocs.io/en/latest/BAC0.core.devices.html#BAC0.core.devices.create_objects.create_object

Doesn't bacnet.create_object create an object on my device?

As opposed to stuff using the ObjectFactory, which creates objects locally.

**Reply by bbartling** [reply]
I guess I am not exactly sure... most of my work is just performing read and write requests in IoT applications where I am just requesting values or writing to someone else devices on a LAN which in the case I am a BACnet client application only. I guess I have never thought of it in the way that a client only BACnet device has "objects".

Else if I want my device to be BACnet discoverable to the network in rare use cases in a simple server application, then yes absolutely my BACnet points are all objects that can represent many different things like in bacpypes library objects.py is very long:
https://github.com/JoelBender/bacpypes/blob/master/py34/bacpypes/object.py

In a client application I can definitely request a million different types of "objects" from a device if the device supports it...

Hopefully this helps

**Reply by tnordstrom** [reply]
Higher level what I want to do is, say I have a device that's been factory reset or equivalent, it only has minimal objects (including one device object), and I'd like to figure out an automated way to create all supported objects on my device. The existing approach is to manually create all these objects with software tools, and it's time consuming. For example, Value objects support 512 objects on some devices, I want to automate creating all 512. I thought I could use create_object for this. Once I'm done creating all these objects, sure then I will request things from them.

**Reply by bbartling** [reply]
Ah okay. That's almost like making a "clone" of a device, right?

So if you were on a LAN with this device that has 512 points you would need to discover it and then do a point discovery process and loop through every BACnet object attribute... 

You know what I would almost recommend is try using bacpypes snapshot:
https://github.com/JoelBender/bacpypes-snapshot

Its a tool that just snaps shots every possible point attribute on a BACnet device. You need to know a few things on bacpypes which is a few notches harder IMHO than BAC0 but its a pretty cool tool.

The output is an SQL lite database that you can replay through Python everything that was snapshotted... 

It wont create a clone BACnet device if thats what your looking for... and if thats the case I would hard code and manually replicate your 512 points which can be done it would look something similar to the first link I sent where the BACnet server is created in a loop.
https://github.com/ChristianTremblay/BAC0/issues/381#issue-1607475583

**Reply by tnordstrom** [reply]
I'm not trying to clone anything.
My device supports objects.
My device's objects can be deleted or created.
A "brand new" device has a minimal set of objects.
Devices support a maximum number of each type of object.
I want to write a Python script that creates this maximum number of objects of all supported types.

I'm not interested in a snap shot of all my attributes. I just want to create objects.

Question: does call create_object do what I need? The documentation for this call is virtually nonexistent. No examples, no description.

**Reply by tnordstrom** [reply]
Section 15.3 of the BACnet standard goes over CreateObject service @ChristianTremblay maybe you could explain the create_object method a little?

**Reply by ChristianTremblay** [reply]
It is a clash of name. 

Let say you have a bacnet device provided by manufacturer X. 

Normally, there is a proprietary software used to configure/program that device that is provided by the manufacturer. 

Of all the devices I saw in the field, I never saw one supporting the dynamic creation of object through bacnet. 

The function inside BAC0 isn't meant to be a bacnet request to dynamically create an object in a device. It was meant as a helper to create local objects in the BAC0 instance.

**Reply by tnordstrom** [reply]
Oh, ok. Thanks for the clarification.
